music/src/core/melodies.ts

```diff
@@ -0,0 +1,142 @@
+/**
+ * Module containing functions for converting to and from quantized melodies.
+ *
+ * @license
+ * Copyright 2020 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Imports
+ */
+import {INoteSequence, NoteSequence} from '../protobuf/index';
+
+import * as sequences from './sequences';
+
+export const NO_EVENT = 0;
+export const NOTE_OFF = 1;
+const FIRST_PITCH = 2;
+
+/**
+ * Melody representation with note onsets, `NO_EVENT` (sustain) and `NOTE_OFF`.
+ *
+ * @param events An array of melody events.
+ * @param minPitch The minimum pitch to represent.
+ * @param maxPitch The maximum pitch to represent.
+ *
+ */
+export class Melody {
+  readonly events: Int32Array;
+  readonly minPitch: number;
+  readonly maxPitch: number;
+
+  constructor(events: Int32Array, minPitch: number, maxPitch: number) {
+    this.events = events;
+    this.minPitch = minPitch;
+    this.maxPitch = maxPitch;
+  }
+
+  /**
+   * Extract a melody from a `NoteSequence`.
+   *
+   * @param noteSequence `NoteSequence` from which to extract a melody.
+   * @param minPitch The minimum pitch to represent. Those above this value will
+   * cause an error to be thrown.
+   * @param maxPitch The maximum pitch to represent. Those above this value will
+   * cause an error to be thrown.
+   * @param ignorePolpyhony If false, an error will be raised when notes start
+   * at the same step. If true, the highest pitched note is used and others are
+   * ignored.
+   * @param numSteps The length of each sequence.
+   * @returns A `Melody` created from the `NoteSequence`.
+   */
+  static fromNoteSequence(
+      noteSequence: INoteSequence, minPitch: number, maxPitch: number,
+      ignorePolyphony = false, numSteps?: number) {
+    sequences.assertIsQuantizedSequence(noteSequence);
+    // Sort by note start times, and secondarily by pitch descending.
+    const sortedNotes: NoteSequence.INote[] =
+        noteSequence.notes.sort((n1, n2) => {
+          if (n1.quantizedStartStep === n2.quantizedStartStep) {
+            return n2.pitch - n1.pitch;
+          }
+          return n1.quantizedStartStep - n2.quantizedStartStep;
+        });
+    const events = new Int32Array(numSteps || noteSequence.totalQuantizedSteps);
+    let lastStart = -1;
+    sortedNotes.forEach(n => {
+      if (n.quantizedStartStep === lastStart) {
+        if (!ignorePolyphony) {
```

Comment:
I would add this as a helper in sequences, like `assertIsMonophonic`

Reply:
(just how you have `assertIsQuantized` above

Reply:
One issue is that isQuantized is a constant time check whereas isMonophonic requires checking the start times of all the notes, so here (I just moved this code from data.ts) it's done in the already-existing melody conversion loop.

Reply:
ohhh i see, yeah got it.